tests/lib/fakestore/store/store.go

```diff
@@ -539,106 +648,216 @@ type detailsResultV2 struct {
+				if err != nil {
+					http.Error(w, fmt.Sprintf("internal error searching assertion backend: %v", err), 500)
+					return
+				}
+
+				// get the current revision from the context for the refresh action
+				currentRevision := 0
+				found := false
+				for _, snapContext := range reqData.Context {
+					if snapContext.SnapID == snapID {
+						currentRevision = snapContext.Revision
+						found = true
+						break
+					}
+				}
+				if !found {
+					http.Error(w, fmt.Sprintf("missing context for snap-id %s in request", snapID), 400)
+					return
+				}
+
+				found = false
+				newestRev := &asserts.SnapRevision{}
+				for _, revAssert := range foundRevisions {
+					snapRevAssert := revAssert.(*asserts.SnapRevision)
+					if snapRevAssert.SnapRevision() > currentRevision {
+						newestRev = snapRevAssert
+						found = true
+					}
+				}
+				if found {
+					// TODO: maybe instead we should use the sha3-384 eventually
+					//       instead of snap files named like core20_12.snap to
+					//       identify the snap revision, but this would require
+					//       naming files with their sha3-384 which is a bit of a
+					//       pain
+
+					if fn, ok := snaps[snapRevRef{revision: newestRev.SnapRevision(), name: name}]; ok {
+						essInfo, err := snapEssentialInfo(w, fn, snapID, bs)
+						if essInfo == nil {
+							if err != errInfo {
+								panic(err)
+							}
+							return
+						}
+
+						// use req.Host here in case the fakestore is being used to
+						// serve snaps to something like a VM, which will be
+						// configured to talk to localhost the way some of the
+						// spread tests are
+						downloadURL := fmt.Sprintf("http://%s", req.Host)
+
+						res := &snapActionResult{
+							Result:      a.Action,
+							InstanceKey: a.InstanceKey,
+							SnapID:      essInfo.SnapID,
+							Name:        essInfo.Name,
+							Snap: detailsResultV2{
+								Architectures: []string{"all"},
+								SnapID:        essInfo.SnapID,
+								Name:          essInfo.Name,
+								Version:       essInfo.Version,
+								Revision:      essInfo.Revision,
+								Confinement:   essInfo.Confinement,
+								Type:          essInfo.Type,
+							},
+						}
+						res.Snap.Publisher.ID = essInfo.DeveloperID
+						res.Snap.Publisher.Username = essInfo.DevelName
+						res.Snap.Download.URL = fmt.Sprintf("%s/download/%s", downloadURL, filepath.Base(fn))
+						res.Snap.Download.Sha3_384 = hexify(essInfo.Digest)
+						res.Snap.Download.Size = essInfo.Size
+						replyData.Results = append(replyData.Results, res)
+
+					} else {
+						http.Error(w, fmt.Sprintf("internal error finding snap file for snap %s revision %d", name, newestRev.Revision()), 500)
+						return
+					}
+				}
+
+				// TODO: what should happen in the else case? probably nothing? it
+				//       just means we didn't find any newer snap revisions for this
+				//       snap, but unclear what the real store returns in this
+				//       situation
```

Comment:
the real store just return the current tip revision of the channel, basically the naive thing that we should do is just find the highest revision and always return it. Snapd has its own logic to ignore it if it's the same as the current revision. The contextual info is used for gating and epoch logic, not to decide if there's a newer revision or not.